content/blog/semgrep-post-message/index.md

```diff
@@ -0,0 +1,175 @@
+
+Before we start writing our new rule it helps to document the patterns we want to match. Besides having a clear spec to work with it also makes it easier to reason about which operators we will need.
+
+## Patterns to flag
+
+```js
+// Inline function without origin check
+window.addEventListener("message", function(evt){
+  console.log('No origin check!');
+});
+
+// External function without origin check
+function receiveMessage(evt) {
+  console.log('No origin check!');
+}
+
+window.addEventListener("message", receiveMessage, false);
+
+// Inline arrow function without origin check
+window.addEventListener('message', (evt) => {
+  console.log('No origin check!');
+});
+
+// External arrow function without origin check
+const arrowHandler = (evt) => {
+    console.log('No origin check!');
+};
+
+window.addEventListener("message", arrowHandler, false);
+```
+
+Knowing that we need to match `inline` or `external` declarations gives us an idea that we will need to rely on the [patterns-either](https://semgrep.dev/docs/writing-rules/rule-syntax/#pattern-either) operator. 
+
+We might be tempted to write four different patterns for this scenario, but since Semgrep is semantic we only need to write **two**! This means that patterns matching and old `function` declaration will also match code using `arrow functions`. 
+
+It's worth repeating this:
+
+> Since Semgrep is semantic, patterns matching and old `function` declaration will also match code using `arrow functions`
+
+ Now, which patterns do we consider safe for origin checking?
+
+## Safe patterns
+
+If the function specified above contains any of the following patterns we **should not** flag them:
+
+```js
+if (evt.origin == "http://example.com") { ... }
+
+if (evt.origin === "http://example.com") { ... }
+
+if (evt.origin != "http://example.com") { ... }
+
+if (evt.origin !== "http://example.com") { ... }
+
+if (someRegex.test(evt.origin)) { ... }
+```
+<br />
+We have our spec, we know what to flag and more importantly, what not to. Time to start working on our rule!
+
+## Writing our rule
+
+Let's take care of the `inline` functions first since they don't need external context:
+
+#### Inline functions
+
+```yaml
+patterns:
+- pattern: |
+  window.addEventListener('message', $FUNC, ...)
+```
+
+`$FUNC` will capture the inline function into a [metavariable](https://semgrep.dev/docs/writing-rules/pattern-syntax/#metavariables). Now we can use the handy [metavariable-pattern](https://semgrep.dev/docs/writing-rules/rule-syntax/#metavariable-pattern) to exclude our **safe patterns**.
+
+```yaml
+patterns:
+- pattern: |
+    window.addEventListener('message', $FUNC, ...)
+- metavariable-pattern:
+    metavariable: $FUNC
+    patterns:
+      - pattern: |
+          function($OBJ) { ... }
+      - pattern-not: |
```

Comment:
> However, the downside of doing it this way is that any reference to $OBJ.origin in the if statement will match, even if it doesn't perform the sort of validation that makes it safe.

I've struggled with this decision to be honest, as a compromise I will list this as an alternative. If nothing else it will teach people about this neat feature. 😄